internal/pubsub/subscriber.go

```diff
@@ -103,167 +90,250 @@ func (s Subscriber) Listen() error {
+func (s *Subscriber) handleDeleteEvent(ctx context.Context, msg events.Request[events.AuthRelationshipRequest, events.AuthRelationshipResponse]) error {
+	elogger := s.logger.With(
+		"event.message.topic", msg.Topic(),
+		"event.message.action", msg.Message().Action,
+		"event.message.object.id", msg.Message().ObjectID.String(),
+		"event.message.relations", len(msg.Message().Relations),
+	)
+
+	var errors []error
+
+	if err := msg.Message().Validate(); err != nil {
+		errors = multierr.Errors(err)
+	}
+
+	resource, err := s.qe.NewResourceFromID(msg.Message().ObjectID)
 	if err != nil {
-		s.logger.Errorw("error deleting relationships - will not reprocess", "error", err.Error())
+		elogger.Warnw("error parsing resource ID", "error", err.Error())
+
+		errors = append(errors, err)
 	}
 
-	return nil
-}
+	rType := s.qe.GetResourceType(resource.Type)
+	if rType == nil {
+		elogger.Warnw("error finding resource type", "error", err.Error())
 
-func (s *Subscriber) handleCreateEvent(ctx context.Context, msg *message.Message, changeMsg events.ChangeMessage) error {
-	resource, err := s.qe.NewResourceFromID(changeMsg.SubjectID)
-	if err != nil {
-		s.logger.Warnw("error parsing subject ID - will not reprocess", "event_type", changeMsg.EventType, "error", err.Error())
+		errors = append(errors, fmt.Errorf("%w: resource: %s", ErrUnknownResourceType, resource.Type))
+	}
 
-		return nil
+	relationships := make([]types.Relationship, len(msg.Message().Relations))
+
+	for i, relation := range msg.Message().Relations {
+		subject, err := s.qe.NewResourceFromID(relation.SubjectID)
+		if err != nil {
+			elogger.Warnw("error parsing subject ID", "error", err.Error())
+
+			errors = append(errors, fmt.Errorf("%w: relation %d", err, i))
+
+			continue
+		}
+
+		sType := s.qe.GetResourceType(subject.Type)
+		if sType == nil {
+			elogger.Warnw("error finding subject resource type", "error", err.Error())
+
+			errors = append(errors, fmt.Errorf("%w: relation %d subject: %s", ErrUnknownResourceType, i, subject.Type))
+
+			continue
+		}
+
+		relationships[i] = types.Relationship{
+			Resource: resource,
+			Relation: relation.Relation,
+			Subject:  subject,
+		}
 	}
 
-	return s.createRelationships(ctx, msg, resource, changeMsg.AdditionalSubjectIDs)
+	if len(errors) != 0 {
+		respondRequest(ctx, elogger, msg, errors...)
+	}
+
+	err = s.deleteRelationships(ctx, relationships)
+
+	respondRequest(ctx, elogger, msg, err)
+
+	return nil
 }
 
-func (s *Subscriber) handleDeleteEvent(ctx context.Context, msg *message.Message, changeMsg events.ChangeMessage) error {
-	resource, err := s.qe.NewResourceFromID(changeMsg.SubjectID)
-	if err != nil {
-		s.logger.Warnw("error parsing subject ID - will not reprocess", "event_type", changeMsg.EventType, "error", err.Error())
+func respondRequest(ctx context.Context, logger *zap.SugaredLogger, msg events.Request[events.AuthRelationshipRequest, events.AuthRelationshipResponse], errors ...error) {
```

Comment:
Two things:
* Adding a span here would be good
* This should probably return an error so we know if we failed to respond

Reply:
updated to start a span for pubsub.respond which sets an error if there's an issue responding to the request. this reply error is then returned resulting in a nak and will requeue and retry.